### PR TITLE
Define `.ty()` for expr, typechecking for applications of callableIdents

### PR DESCRIPTION
**Pull request by zgrannan**
This PR also pulls in changes from https://github.com/Aurel300/prusti-dev/pull/26, https://github.com/Aurel300/prusti-dev/pull/34, https://github.com/Aurel300/prusti-dev/pull/35 (I suppose in theory this could be merged and the others closed).

Defines getting types for expressions, and type-checking of function application. Since we are using viper generics in some places, this also includes logic for performing substitutions (in particular, the return type of a function application may have substitutions applied).

### CODE REVIEW

File: prusti-encoder/src/encoders/mir_pure_function.rs [hunk 1]
```diff
@@ -79,8 +79,9 @@ impl TaskEncoder for MirFunctionEnc {
                 .map(|def_idx| local_defs.locals[def_idx].ty.snapshot)
                 .collect();
             let args = UnknownArity::new(vcx.alloc_slice(&args));
-            let function_ref = FunctionIdent::new(function_name, args);
-            deps.emit_output_ref::<Self>(*task_key, MirFunctionEncOutputRef { function_ref, return_type: local_defs.locals[mir::RETURN_PLACE].ty });
+            let return_type = local_defs.locals[mir::RETURN_PLACE].ty;
+            let function_ref = FunctionIdent::new(function_name, args, return_type.snapshot);
+            deps.emit_output_ref::<Self>(*task_key, MirFunctionEncOutputRef { function_ref, return_type });
```
Comment (Aurel300) [inline prusti-encoder/src/encoders/mir_pure_function.rs]: Is the `return_type` field of this output ref obsolete now?

File: vir/src/gendata.rs [hunk 2]
```diff
@@ -117,7 +132,34 @@ pub enum ExprKindGenData<'vir, Curr: 'vir, Next: 'vir> {
...
 }
+
+impl<'vir, Curr, Next> ExprKindGenData<'vir, Curr, Next> {
+    pub fn ty(&self) -> Type<'vir> {
+        match self {
+            ExprKindGenData::Local(l) => l.ty,
+            ExprKindGenData::Field(_, f) => f.ty,
+            ExprKindGenData::Old(e) => e.ty(),
+            ExprKindGenData::Const(c) => c.ty(),
+            ExprKindGenData::Result => &TypeData::Ref, // For now this always seems to be the case
```
Comment (Aurel300) [inline vir/src/gendata.rs]: This seems wrong. `Result` should be the node referring to the value returned by a pure function, which in general is a snapshot.

File: vir/src/gendata.rs [hunk 3]
```diff
@@ -117,7 +132,34 @@ pub enum ExprKindGenData<'vir, Curr: 'vir, Next: 'vir> {
...
+            ExprKindGenData::AccField(_) => &TypeData::Bool,
+            ExprKindGenData::Unfolding(f) => f.expr.ty(),
+            ExprKindGenData::UnOp(u) => u.expr.ty(), // For now this always seems to be the case
```
Comment (Aurel300) [inline vir/src/gendata.rs]: What would be a unary op that does not behave this way?

File: prusti-encoder/src/encoders/mir_pure_function.rs (hunk 1 above)
Comment (zgrannan) [reply prusti-encoder/src/encoders/mir_pure_function.rs]: Yeah it doesn't seem necessary, removed it.

File: vir/src/gendata.rs (hunk 2 above)
Comment (zgrannan) [reply vir/src/gendata.rs]: It looks like there are also pure functions returning `Ref`s, for example the function that gets a struct field ref from a struct ref.

I fixed this by changing `Result` to also hold a type; which should allow handling any case.

File: vir/src/gendata.rs (hunk 3 above)
Comment (zgrannan) [reply vir/src/gendata.rs]: Actually I think that in Viper all of the ops do behave that way, so I removed the comment.

**Comment by zgrannan** [comment]
@Aurel300 updated based on your comments!

**Comment by Aurel300** [comment]
I'll merge this now, but I think in a future PR we should have unit tests checking e.g. `check_types`. There are some edge cases I can think of (domains with the same type parameter name?) that might not pop up in the small test file(s) we have now.
